transformer_engine/common/comm_gemm_overlap/userbuffers/userbuffers.cu

```diff
@@ -2597,30 +2598,54 @@ void reset_counters(void *atomic_ptr, int num_chunks, bool allgather, cudaStream
   reset_counters_kernel<<<grid, block, 0, stream>>>(atomic_ptr, num_chunks, allgather);
 }
 
-template <typename fp8type>
+template <typename fp8type, int nvec>
 __global__ void __launch_bounds__(MAX_THREADS / 4)
     reduce_fp8_in_bf16_out_cuda(void *inputs, void *output, const float *scale,
-                                const int num_inputs, const int input_size) {
-  const size_t tid = threadIdx.x + blockDim.x * blockIdx.x;
+                                const int num_inputs, const int input_size,
+                                const int num_aligned_elements_per_input,
+                                const int tot_input_size) {
   fp8type *inputs_fp8 = reinterpret_cast<fp8type *>(inputs);
-  float accum_buf = static_cast<float>(inputs_fp8[tid]) * (*scale);
+  half_dtype *output_half = reinterpret_cast<half_dtype *>(output);
+
+  transformer_engine::VectorizedLoader<fp8type, nvec, true> loader(inputs_fp8, tot_input_size);
+  transformer_engine::VectorizedStorer<half_dtype, nvec, true> storer(output_half, input_size);
+
+  const size_t tid = threadIdx.x + blockDim.x * blockIdx.x;
+  float accum_buf[nvec];
+
+  loader.load(tid, tot_input_size);
 #pragma unroll
-  for (int i = 1; i < num_inputs; i++) {
-    accum_buf += static_cast<float>(inputs_fp8[tid + input_size * i]) * (*scale);
+  for (int i = 0; i < nvec; ++i) {
+    accum_buf[i] = static_cast<float>(loader.separate()[i]) * (*scale);
   }
-  half *output_half = reinterpret_cast<half *>(output);
-  output_half[tid] = (half)accum_buf;
+#pragma unroll
+  for (int input_id = 1; input_id < num_inputs; ++input_id) {
+    loader.load(tid + num_aligned_elements_per_input * input_id, tot_input_size);
+#pragma unroll
+    for (int i = 0; i < nvec; ++i) {
+      accum_buf[i] += static_cast<float>(loader.separate()[i]) * (*scale);
+      if (input_id == num_inputs - 1) {
+        storer.separate()[i] = static_cast<half_dtype>(accum_buf[i]);
+      }
+    }
+  }
```

Comment:
- Correctness bug when `num_inputs == 1`
- Unrolling loop over `num_inputs` is not necessary since it's not known at compile-time
```suggestion
  for (int input_id = 1; input_id < num_inputs; ++input_id) {
    loader.load(tid + num_aligned_elements_per_input * input_id, tot_input_size);
#pragma unroll
    for (int i = 0; i < nvec; ++i) {
      accum_buf[i] += static_cast<float>(loader.separate()[i]) * (*scale);
    }
  }
#pragma unroll
  for (int i = 0; i < nvec; ++i) {
    storer.separate()[i] = static_cast<half_dtype>(accum_buf[i]);
  }
```
Same issue in `reduce_bf16_cuda`.